PSA: use static key slots to store keys

valeriosetti opened the pull request:
## Description

This PR proposes an alternative solution for key slots which preallocates all the required buffers instead of dynamically allocating them at runtime. A new build symbol named `MBEDTLS_PSA_STATIC_KEY_SLOTS` is used for this purpose and it's disabled by default so that it cannot affect normal Mbed TLS users.

This change helps removing the code required to manage heap memory in embedded applications (assuming no other component is using heap, of course) and therefore reducing ROM footprint. This comes with a slightly higher RAM usage, of course, but this can optionally be fine tuned using `MBEDTLS_PSA_KEY_SLOT_COUNT`.

## Background

`p256-m` driver was recently added to Zephyr as alternative to TinyCrypt for secp256r1. Albeit `p256-m` [claimed to have a smaller footprint](https://github.com/mpg/p256-m) compared to TinyCrypt results showed that this does not hold if standard PSA interface provided by Mbed TLS is used to access this driver.
Here's a comparison of an ad-hoc test program that was added to Zephyr performing the same operations on secp256r1 curve (2x key derivation + 1x key agreement) with Mbed TLS
```
west build -p always -b nrf52840dk/nrf52840 -T tests/crypto/secp256r1/crypto.secp256r1.mbedtls
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       44363 B         1 MB      4.23%
             RAM:       15680 B       256 KB      5.98%
        IDT_LIST:          0 GB        32 KB      0.00%
```
and TinyCrypt
```
west build -p always -b nrf52840dk/nrf52840 -T tests/crypto/secp256r1/crypto.secp256r1.tinycrypt
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       40707 B         1 MB      3.88%
             RAM:         14 KB       256 KB      5.47%
        IDT_LIST:          0 GB        32 KB      0.00%
```
Digging a bit into the generated binaries it turned out that Mbed TLS had a larger footprint than TinyCrypt because of:
- some code overhead in PSA core between the call to PSA APIs and the actual entry of `p256-m` driver;
- key management.

The latter is not present in TinyCrypt and most of it (around 2kB) is due to heap management code enabled in the Zephyr core.

## Results

With this patch in place it is possible to get rid of extra heap management code (assuming no one else is using it in the program, of course). Here's the same example used before built with this patch:

```
west build -p always -b nrf52840dk/nrf52840 -T tests/crypto/secp256r1/crypto.secp256r1.mbedtls
...
Memory region         Used Size  Region Size  %age Used
           FLASH:       41935 B         1 MB      4.00%
             RAM:       17856 B       256 KB      6.81%
        IDT_LIST:          0 GB        32 KB      0.00%
```
There's clearly a 2.4 kB ROM code reduction at the expense of a larger RAM usage. However RAM usage can be further reduced by overriding default `MBEDTLS_PSA_KEY_SLOT_COUNT` if the developer knows a-priori the amount of keys that are needed for the application.

## Dependencies

This PR depends on:

- https://github.com/Mbed-TLS/mbedtls-framework/pull/48

This PR happens to resolve #9577 (that is not an issue in 3.6).

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **development** here
- [x] **3.6 backport**  https://github.com/Mbed-TLS/mbedtls/pull/9448
- [x] **2.28 backport** not required because new feature
- [x] **framework** https://github.com/Mbed-TLS/mbedtls-framework/pull/48
- [x] **tests** provided

gilles-peskine-arm commented on library/psa_crypto_core.h:
```diff
@@ -55,6 +55,20 @@ typedef enum {
+#define MAX3(a, b, c) ((a) < (b) ? (b) < (c) ? (c) : (b) : (a) < (c) ? (c) : (a))
+
+#define KEY_SLOT_BUFFER_SIZE        \
+        MAX3(PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE, PSA_HASH_MAX_SIZE, PSA_CIPHER_MAX_KEY_LENGTH)
```
This does not account for all key types. In practice, on devices that want to avoid dynamic allocation, it's likely that public keys will be the largest thing unless asymmetric cryptography is completely disabled, and hashes otherwise. But that's still an arbitrary limitation that would need to be at the very least documented, and perhaps we should give a way to configure it.

valeriosetti replied:
> This does not account for all key types

Yeah, I was aware of this limitation. For the list of supported key types I get it from `crypto_values.h` and I think I've included support for `HMAC, AES, ARIA, DES, CAMELLIA, CHACHA20, RSA_PUBLIC_KEY, RSA_KEY_PAIR, ECC_PUBLIC_KEY_BASE, ECC_KEY_PAIR_BASE, DH_PUBLIC_KEY_BASE, DH_KEY_PAIR_BASE`.
I skipped RAW_DATA` because IMO it didn't make much sense for what I was trying to achieve, while for `DERIVE, PASSWORD, PASSWORD_HASH, PEPPER` I didn't know exactly what to set. 
Do you think it's worth to add support also for the missing key types? Or perhaps we can just move to the 2nd point (because it has a possible alternative to this)

> and perhaps we should give a way to configure it.

Do you mean that I should add another build symbol to set the max slot size manually? In this case it can have a default value (for example `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE`), but then the user can change it at build time if he knows that he's going to use a larger key for some reason. This would also greatly simplify the previous part of this comment and allow to include also `RAW_DATA` key type. Is my understanding correct?

gilles-peskine-arm replied:
There's no intrinsic upper limit for things like `RAW_DATA` and `DERIVE`. In practice, `DERIVE` is usually 128 or 256 bits for root secrets and intermediate keys, and the size of the largest elliptic curves for secrets from key agreement. `RAW_DATA` could be anything, for example it's often used to store a certificate in a secure enclave alongside the corresponding private key. So yes, for practicality, a numerical setting that defaults to `PSA_EXPORT_PUBLIC_KEY_MAX_SIZE` is probably the right thing.

gilles-peskine-arm replied:
Another case is wrapped keys stored on behalf of a secure element. These have the size of the key plus some overhead, typically around 32 to 64 bytes.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.data:
```diff
@@ -35,7 +35,7 @@ parse_storage_data_check:"505341004b45590000000000010000000170000001000000000000
 
 # Not specific to files, but only run this test in an environment where the maximum size could be reached.
 Save maximum-size persistent raw key
```
When `MBEDTLS_PSA_STATIC_KEY_SLOTS` is enabled, we should have a similar test case with the maximum in-memory key size plus the storage overhead.

gilles-peskine-arm commented on tf-psa-crypto/include/psa/crypto_sizes.h:
```diff
@@ -146,8 +146,10 @@
-#else /* SHA-1 or smaller */
+#elif defined(PSA_WANT_ALG_SHA_1)
 #define PSA_HASH_MAX_SIZE 20u
+#else /* None of the above defined */
```
This is not correct for a configuration with only MD4 or RIPEMD160. Such a configuration is unlikely, but not impossible.

gilles-peskine-arm replied:
Sorry, I meant MD5, not MD4.

gilles-peskine-arm commented on tf-psa-crypto/include/psa/crypto_sizes.h:
```diff
@@ -1085,6 +1091,23 @@
+ */
+#if (defined(PSA_WANT_KEY_TYPE_AES) || defined(PSA_WANT_KEY_TYPE_ARIA) || \
+     defined(PSA_WANT_KEY_TYPE_CAMELLIA) || defined(PSA_WANT_KEY_TYPE_CHACHA20))
+#define PSA_CIPHER_MAX_KEY_LENGTH       32u
```
We should at least make a note that this size needs to double if `PSA_WANT_ALG_XTS` (currently we don't support XTS through PSA).

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -1246,6 +1246,14 @@ component_test_default_cmake_gcc_asan_new_bignum () {
 
+component_test_crypto_with_static_key_slots() {
+    msg "build: crypto config + MBEDTLS_PSA_STATIC_KEY_SLOTS"
+    scripts/config.py crypto
```
At least
```suggestion
    scripts/config.py crypto_full
```

We need to validate compatibility with `MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS`, at least. Probably other things I'm not thinking of to start with.

gilles-peskine-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -3206,6 +3206,24 @@
+ * Requires: MBEDTLS_PSA_CRYPTO_C
+ *
+ */
+//#define MBEDTLS_PSA_STATIC_KEY_SLOTS
```
I'm not sure if it's technically possible to have both this and [dynamic key slots](https://github.com/Mbed-TLS/mbedtls/pull/9240), but we should exclude it because it would be additional testing burden and it would make no sense.

We also need to make the naming non-confusing, since there will be three states: dynamically allocated key slices (for the A-class world), statically allocated key slices but dynamically allocated key material (for when RAM and code size are limited but there's still a heap), and fully static (for going without a heap).

valeriosetti replied:
> I'm not sure if it's technically possible to have both this and https://github.com/Mbed-TLS/mbedtls/pull/9240, but we should exclude it because it would be additional testing burden and it would make no sense.

Oh, I wasn't aware of https://github.com/Mbed-TLS/mbedtls/pull/9240, but I agree that the build option/feature added there is in contrast to the one added here, so it makes no sense to enabled them both at the same time. Since this PR need a design approval while yours is on review, I guess that 9240 will likely be merged before this one. Once that happens I will rebase this PR on top of `development` and do the proper changes.

> We also need to make the naming non-confusing, since there will be three states

No problem at all on this, I just pick something that made sense to me, but I'm completely open to new/better ideas.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -1246,6 +1246,14 @@ component_test_default_cmake_gcc_asan_new_bignum () {
     tests/context-info.sh
 }
 
+component_test_crypto_with_static_key_slots() {
```
If keystore without malloc is a feature, we should make a build where only libtestdriver1 and test code has access to malloc, but the `mbedtls_calloc` used to build the core always fails. (Of course no md/cipher/pk/x509/tls in that build.)

valeriosetti replied:
I think I need a bit of help on this. The [last commit](https://github.com/Mbed-TLS/mbedtls/pull/9302/commits/b38c62fefd40d7584bf6aa0b9e2f9d538ddb2c8f) I added tries to implement what (I think) you asked in this comment, i.e. a test component where `libtestdriver1` and `test_suite_xxx` can use `malloc/free` whereas the library (including PSA core) cannot. First question: albeit this is a rough implementation is that close to what you had in mind?

gilles-peskine-arm replied:
This looks on the right track: compiling the core with different settings for malloc/free compared with libtestdriver1 and the test code. It's incomplete because with `MBEDTLS_PLATFORM_MEMORY` undefined, `mbedtls_calloc` is just `#define`d to `calloc`. We could grep `libmbedcrypto.a` for `calloc`, or arrange to define `mbedtls_calloc` to `if_you_get_a_link_error_this_test_is_failing`, or any other solution that results in the component failing if the core calls `mbedtls_calloc`.

valeriosetti replied:
> or arrange to define mbedtls_calloc to if_you_get_a_link_error_this_test_is_failing

I found this to be the simplest solution. Unfortunately test component is still failing in https://github.com/Mbed-TLS/mbedtls/pull/9302/commits/079ad6c883cf072163f471d87e4fe9b97048c04b, but I'm working on that to fix it.

valeriosetti replied:
As of https://github.com/Mbed-TLS/mbedtls/pull/9302/commits/c22f65211e8b6c686a71de41833243ba66c2b78f this PR should have the CI green and therefore it should be ready for review ;)

gilles-peskine-arm reviewed (changes requested):
This is a nontrivial feature and we need to discuss both prioritization and architecture further. A malloc-less subset of Mbed TLS is something we've discussed for a while but so far decided not to spend resources on it.

Technically this is a new feature, and we don't normally add features to an LTS. I'm ok with that because it's gated by a compile-time option, but this needs product management approval.

The scheduling needs to take https://github.com/Mbed-TLS/mbedtls/pull/9240 into account, which is a MUST for Mbed TLS 3.6.1.

gilles-peskine-arm commented:
As the key store design is getting a bit complex, I've written a design document in https://github.com/Mbed-TLS/mbedtls/pull/9309 . It's aligned with this pull request (if not, I made a mistake somewhere, please let me know), except that I've documented `MBEDTLS_PSA_KEY_SLOT_BUFFER_SIZE` as a publicly configurable option [as discussed above](https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1648543938).

gilles-peskine-arm commented:
I'm formally approving the design: a malloc-free subset of the library is consistent with our long-term objective, and it's nice that we can have it earlier than planned.

However, we're still discussing internally whether the numbers on Zephyr justify backporting to the 3.6 LTS branch.

valeriosetti commented:
Only InternalCI reported a failure, while the OpenCI is completely fine. I think this is ready for reviews again :)

gilles-peskine-arm commented on tf-psa-crypto/core/psa_crypto.c:
```diff
@@ -7965,7 +7988,13 @@ psa_status_t psa_generate_key_ext(const psa_key_attributes_t *attributes,
+    int is_slot_unused = (slot->key.data == NULL);
+#endif
+
+    if (is_slot_unused) {
```
As in `psa_allocate_buffer_to_slot` above, I'm pretty sure this test is always true and I've checked that the else path is not triggered by our test suite in the default configuration.

gilles-peskine-arm reviewed (changes requested):
The design looks good to me. I want more test coverage, and a few minor improvements in various places. I don't foresee a conflict with the dynamic key store.

valeriosetti commented on tf-psa-crypto/tests/suites/test_suite_psa_crypto_entropy.function:
```diff
@@ -8,6 +8,10 @@
 #include "entropy_poll.h"
 
 /* Calculating the minimum allowed entropy size in bytes */
+#if !defined(MBEDTLS_ENTROPY_BLOCK_SIZE)
+#define MBEDTLS_ENTROPY_BLOCK_SIZE      MBEDTLS_ENTROPY_MIN_PLATFORM
+#endif
```
Actually  `check_config.h` has a check for [this condition](https://github.com/Mbed-TLS/mbedtls/blob/b8457fff9fd4303b2dfd91d961aed5b81db5e8ef/include/mbedtls/check_config.h#L689) so I suspect that the problem I was trying to fix here was something else. Anyway, since I don't remember where it was exactly and, yes, it was also a quick fix, I can just revert this change and see what happens.

valeriosetti commented on tf-psa-crypto/core/psa_crypto.c:
```diff
@@ -2077,7 +2094,13 @@ psa_status_t psa_import_key(const psa_key_attributes_t *attributes,
+    int is_slot_unused = (slot->key.data == NULL);
+#endif
+
+    if (is_slot_unused) {
```
I actually had issues in `component_test_crypto_with_static_key_slots` for tests with SE driver, so I simply changed `is_slot_unused` with `slot->key.bytes == 0`.

valeriosetti commented on tf-psa-crypto/core/psa_crypto.c:
```diff
@@ -705,6 +705,17 @@ MBEDTLS_STATIC_TESTABLE psa_status_t psa_mac_key_can_do(
+    slot->key.in_use = 1;
+#else
     if (slot->key.data != NULL) {
         return PSA_ERROR_ALREADY_EXISTS;
```
But this check prevents corruption from happening, so I think `PSA_ERROR_ALREADY_EXISTS` makes sense here. Isn't it?

valeriosetti replied:
I saw #9484 and I was also tempted to start looking into it, but then I thought that perhaps waiting for the removal of `MBEDTLS_PSA_CRYPTO_SE_C` could simplify the development so I choose to wait for this to happen. That's planned for 4.0, right?

gilles-peskine-arm replied:
Right. There's a part of #9484 that applies to 3.6, but we might not want to take the risk of changing it in an LTS branch. And another part which will be easier after https://github.com/Mbed-TLS/mbedtls/issues/8151. And we should do https://github.com/Mbed-TLS/mbedtls/issues/9255 before that. They are on the [4.0 board](https://github.com/orgs/Mbed-TLS/projects/15/views/1).

valeriosetti commented:
@gilles-peskine-arm I think I addressed all the comments a part from:
- https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1711629287 which (I suspect) will take some time and
- https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1711938671 which might not be strictly required.

Anyway it would be great if you can have a look at this PR again to check if we can have it backported before the 3.6.1 codefreeze.

valeriosetti commented:
I need a bit of help here for the CI failures as they are all for Windows targets, but I don't have a Win PC to test them. If I'm not wrong the failure is
`273>suites/main_test.function(862): fatal error C1017: invalid integer constant expression [c:\windows\workspace\mbed-tls-pr-head_PR-9302-head\worktrees\tmpcd18d9ao\cmake_solution\tf-psa-crypto\tests\test_suite_psa_crypto.vcxproj]`. This might be connected to `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE<2364` that I added as dependency in `test_suite_psa_crypto.data`, but then I don't understand why Linux test systems are not failing in that case :thinking:

gilles-peskine-arm commented on tf-psa-crypto/tests/suites/test_suite_psa_crypto.data:
```diff
@@ -7481,6 +7481,10 @@ PSA generate key: RSA, maximum size exceeded
 generate_key:PSA_KEY_TYPE_RSA_KEY_PAIR:PSA_VENDOR_RSA_MAX_KEY_BITS+8:PSA_KEY_USAGE_EXPORT | PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT:PSA_ALG_RSA_PKCS1V15_CRYPT:PSA_ERROR_NOT_SUPPORTED:0
 
+PSA generate key: RSA, key pair size does not fit in static key buffer
+depends_on:PSA_WANT_ALG_RSA_PKCS1V15_CRYPT:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_GENERATE:MBEDTLS_PSA_STATIC_KEY_SLOTS:MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE<2364
```
> I need a bit of help here for the CI failures as they are all for Windows targets, but I don't have a Win PC to test them. If I'm not wrong the failure is
`273>suites/main_test.function(862): fatal error C1017: invalid integer constant expression [c:\windows\workspace\mbed-tls-pr-head_PR-9302-head\worktrees\tmpcd18d9ao\cmake_solution\tf-psa-crypto\tests\test_suite_psa_crypto.vcxproj]`. This might be connected to `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE<2364` that I added as dependency in `test_suite_psa_crypto.data`, but then I don't understand why Linux test systems are not failing in that case :thinking: 

Both the fact that this is the only change in this file since a passing CI run, and the line number in the C file, confirm that this line in the culprit. In the C file, the line is 780 lines below `#line 82 "suites/main_test.function"`:

```
#if (MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE<2364)
```

In both the default configuration and the full configuration, `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE` is not defined in public headers, so the definition from `psa_crypto_core.h` (which is included at that point in `test_suite_psa_crypto.c`) takes effect:
```
#define MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE  \
    (PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE > PSA_CIPHER_MAX_KEY_LENGTH) ? \
    PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE : PSA_CIPHER_MAX_KEY_LENGTH
```

On Windows, we test in the `full` configuration minus a few non-PSA-related options. Both in the default configuration and in the full configuration, on Linux, this macro expands to
```
(((9u * (((4096u) / 2u + 1u) / 8u + 5u) + 14u) > ((((8192u) + 7u) / 8u))) ? (9u * (((4096u) / 2u + 1u) / 8u + 5u) + 14u) : ((((8192u) + 7u) / 8u)) > 32u) ? ((9u * (((4096u) / 2u + 1u) / 8u + 5u) + 14u) > ((((8192u) + 7u) / 8u))) ? (9u * (((4096u) / 2u + 1u) / 8u + 5u) + 14u) : ((((8192u) + 7u) / 8u)) : 32u
```

There's nothing that could be problematic, in particular no comma operator (which GCC and Clang tolerate) or cast.

So I'm stumped as well.

tom-cosgrove-arm replied:
Making this change (which isn't right, but shows what the problem is) fixes the build

```
diff --git a/tf-psa-crypto/core/psa_crypto_core.h b/tf-psa-crypto/core/psa_crypto_core.h
index 2a358fc040..a6e1fdf77e 100644
--- a/tf-psa-crypto/core/psa_crypto_core.h
+++ b/tf-psa-crypto/core/psa_crypto_core.h
@@ -59,9 +59,7 @@ typedef enum {
  * set it to the maximum between PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE and
  * PSA_CIPHER_MAX_KEY_LENGTH. */
 #if !defined(MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE)
-#define MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE  \
-    (PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE > PSA_CIPHER_MAX_KEY_LENGTH) ? \
-    PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE : PSA_CIPHER_MAX_KEY_LENGTH
+#define MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE  2000
 #endif /* !MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE*/

 /** The data structure representing a key slot, containing key material
```

tom-cosgrove-arm replied:
(I've deleted my previous comment)

Yes, it's the definition of `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE` in `psa_crypto_core.h` that is the problem, but it's not the ternary operator. Replacing all the macros in that line with constants does make it work... investigating a bit further right now

tom-cosgrove-arm replied:
It compiles if I make this one change

```
diff --git a/tf-psa-crypto/core/psa_crypto_core.h b/tf-psa-crypto/core/psa_crypto_core.h
index 2a358fc040..a0380e4b4d 100644
--- a/tf-psa-crypto/core/psa_crypto_core.h
+++ b/tf-psa-crypto/core/psa_crypto_core.h
@@ -61,7 +61,7 @@ typedef enum {
 #if !defined(MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE)
 #define MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE  \
     (PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE > PSA_CIPHER_MAX_KEY_LENGTH) ? \
-    PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE : PSA_CIPHER_MAX_KEY_LENGTH
+    (PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE) : PSA_CIPHER_MAX_KEY_LENGTH
 #endif /* !MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE*/

 /** The data structure representing a key slot, containing key material
```

gilles-peskine-arm commented on tf-psa-crypto/core/psa_crypto.c:
```diff
@@ -1177,11 +1183,14 @@ static psa_status_t psa_get_and_lock_transparent_key_slot_with_policy(
 {
+#if !defined(MBEDTLS_PSA_STATIC_KEY_SLOTS)
     if (slot->key.data != NULL) {
         mbedtls_zeroize_and_free(slot->key.data, slot->key.bytes);
```
MAJOR: we must wipe the key data even with static key slots. If we don't, the key data is marked as unused, but it's still in memory, exposed to future compromises.

gilles-peskine-arm commented on tf-psa-crypto/tests/suites/test_suite_psa_crypto.data:
```diff
@@ -7481,6 +7481,10 @@ PSA generate key: RSA, maximum size exceeded
 
+PSA generate key: RSA, key pair size does not fit in static key buffer
+depends_on:PSA_WANT_ALG_RSA_PKCS1V15_CRYPT:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_GENERATE:MBEDTLS_PSA_STATIC_KEY_SLOTS:MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE<2364
+generate_key:PSA_KEY_TYPE_RSA_KEY_PAIR:PSA_VENDOR_RSA_MAX_KEY_BITS:PSA_KEY_USAGE_EXPORT:PSA_ALG_RSA_PKCS1V15_CRYPT:PSA_ERROR_NOT_SUPPORTED:0
```
With static key slots and `PSA_VENDOR_RSA_MAX_KEY_BITS` changed to a lower value, this test case might fail. We should use a fixed size here, and add a condition on `PSA_VENDOR_RSA_MAX_KEY_BITS` being larger than that fixed size.

gilles-peskine-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -35,9 +35,13 @@ component_test_crypto_with_static_key_slots() {
+    # Intentionally set MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE slightly smaller
+    # than PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE where the latter would be 2364
+    # bytes for an RSA key pair of 4096 bits.
+    scripts/config.py set MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE 2362
```
The size of an RSA key pair representation is slightly variable for a given bit-size, because the format requires removing leading zeros. Given that the test case tries to generate a random key, if the cutoff is only a few bytes below the maximum size, the test case will sometimes fail randomly. We can take a much lower cutoff here (as well as a lower key size, see my comment on the test case). What we want is, in that build, to have at least one test case that successfully creates an RSA key pair, and another one that returns `NOT_SUPPORTED` when doing the same thing but with a larger bit-size that is less than `PSA_VENDOR_RSA_MAX_KEY_BITS`.

valeriosetti replied:
> The size of an RSA key pair representation is slightly variable for a given bit-size, because the format requires removing leading zeros

Oh, I didn't check the PSA format used to store RSA keys and I assumed it just worked the same way as EC keys (where, if I'm not wrong, the size is always the same and it's given from the bit-size).
Thanks for the notification. I will fix this

gilles-peskine-arm commented on tf-psa-crypto/tests/suites/test_suite_psa_crypto_storage_format.function:
```diff
@@ -293,13 +302,32 @@ void key_storage_read(int lifetime_arg, int type_arg, int bits_arg,
+    /* Create a persistent key which is intentionally larger than the specified
+     * bit size. */
+    if (flags & TEST_FLAG_OVERSIZED_KEY) {
+        TEST_CALLOC(custom_key_data, PSA_BITS_TO_BYTES(bits) + 1);
```
Non-blocker: why do +1 here instead of passing `PSA_BYTES_TO_BITS(MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE + 1)` as the `bits` argument?

gilles-peskine-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -3049,6 +3049,26 @@
 #define MBEDTLS_PSA_ITS_FILE_C
 
+/**
+ * \def MBEDTLS_PSA_STATIC_KEY_SLOTS
```
Please add a check in `check_config.h` to forbid having this together with dynamic key slots. As [previously discussed](https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1648561945), it'll probably work, but it's kind of useless and we don't want the testing burden.

valeriosetti replied:
Done. But of course this adds a test dependency on https://github.com/Mbed-TLS/mbedtls/pull/9464 because now the `component_check_names` complains that `MBEDTLS_PSA_KEY_STORE_DYNAMIC` does not exists. Not a problem of course, more a note for upcoming reviewers which see the CI failing on that component.

gilles-peskine-arm reviewed (changes requested):
A zeroization has gone missing, and a few minor things. Looks good to me otherwise.

valeriosetti commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -3839,6 +3859,19 @@
+ *
+ * Define the size (in bytes) of each static key slot when
+ * MBEDTLS_PSA_STATIC_KEY_SLOTS is set. If not
+ * explicitly defined then it's automatically set to hold the maximum
+ * asymmetric PSA key enabled in the build (through PSA_WANT_xxx symbols).
```
Sorry, symmetric keys were not listed here because they were not used in the first implementation. Then when I changed the code, I forgot to update this part.
As for the content of this paragraph and the "promises" it makes, what if I rephrase it as follows:
```
 * Define the size (in bytes) of each static key buffer when
 * #MBEDTLS_PSA_STATIC_KEY_SLOTS is set. If not
 * explicitly defined then it's automatically guessed from available PSA keys
 * enabled in the build through PSA_WANT_xxx symbols.
 ```
To me it looks more generic because we just mention generic "PSA keys" without specifying the type. Is that enough?

valeriosetti commented:
I should have addressed all the comments so far. Please note that some CI failures are [expected](https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1719819697) related to wrong build symbol definitions. This will be fixed once the [depending PR](https://github.com/Mbed-TLS/mbedtls/pull/9464) is merged

gilles-peskine-arm commented:
I'm still busy and probably won't be able to start re-reviewing until tomorrow. Would you mind rebasing today, now that https://github.com/Mbed-TLS/mbedtls/pull/9464 has been merged? And hopefully get a passing CI.

valeriosetti commented:
There's only a single test failing on the CI: `check_doxygen_warnings`. I've tried to reproduce it locally with Docker images of Ubuntu 16.04 and 20.04, but both are successful. What am I missing?

gilles-peskine-arm commented on tf-psa-crypto/core/psa_crypto_core.h:
```diff
@@ -55,6 +55,15 @@ typedef enum {
     PSA_SLOT_PENDING_DELETION,
 } psa_key_slot_state_t;
 
+/* If the size of static key slots is not explicitly defined by the user, then
+ * set it to the maximum between PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE and
+ * PSA_CIPHER_MAX_KEY_LENGTH. */
+#if !defined(MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE)
+#define MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE  \
+    ((PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE > PSA_CIPHER_MAX_KEY_LENGTH) ? \
+     PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE : PSA_CIPHER_MAX_KEY_LENGTH)
+#endif /* !MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE*/
```
To fix `check_doxygen_warnings`, I think that due to a limitation of how we build the documentation, you have to define `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE` in a public header that is not `mbedtls_config.h`. That's how it's done for other non-boolean options, including `MBEDTLS_PSA_KEY_SLOT_COUNT`.

So please move the fallback definition from `psa_crypto_core.h` to `include/psa/crypto_extra.h`.

valeriosetti commented:
I just rebased to solve a conflict with `tf-psa-crypto/tests/suites/test_suite_pkparse.data`. Unless something strange happened during the rebase, the PR has CI green before so I think it's still worth a review.

valeriosetti commented:
I just solved a (new) conflict with `config.py` being partially moved to the framework sub-repo.
Therefore there's now this PR depends on https://github.com/Mbed-TLS/mbedtls-framework/pull/48.

valeriosetti commented:
OpenCI fully green so the PR is ready for another round of reviews :)

gilles-peskine-arm reviewed (changes requested):
Sorry for the delay. This looks good to me except for some of the guards on test cases.

valeriosetti commented on tf-psa-crypto/tests/suites/test_suite_psa_crypto.data:
```diff
@@ -7626,11 +7633,11 @@ depends_on:MBEDTLS_THREADING_PTHREAD
 
 PSA concurrent key generation: raw data, (MBEDTLS_CTR_DRBG_MAX_REQUEST + 1) * 8 bits
-depends_on:MBEDTLS_THREADING_PTHREAD
+depends_on:MBEDTLS_THREADING_PTHREAD:!MBEDTLS_PSA_STATIC_KEY_SLOTS
```
It's been a while since I made this change so I don't remember exactly what it was the problem there (you are totally right: no documentation on this means that even I cannot clearly explain why I did this, sorry). Guessing from the description of the test I would say that the problem here was that these tests are trying to generate a large `PSA_KEY_TYPE_RAW_DATA` and this is unlikely to be successful when static key slot buffers are used.
First of all I will verify that my hypothesis is correct, then I will replace the `depends_on` from `!MBEDTLS_PSA_STATIC_KEY_SLOTS` to something like `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE>x`

valeriosetti replied:
Question: I was thinking to improve this by adding the following `depends_on` to the `PSA concurrent key generation: raw data, (MBEDTLS_CTR_DRBG_MAX_REQUEST + 1) * 8 bits` test:

`MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE > MBEDTLS_CTR_DRBG_MAX_REQUEST + 1`.

However `generate_test_code.py` complains that this is an `Invalid dependency`. I suspect that this might be connected to symbols' expansion. Do you have any suggestion on how to implement this?

valeriosetti replied:
Question: checking `generate_test_code.py` it seems to me that conditional `depends_on` should have the form: 
`<C_symbol> <operator> <numerical value>`. 
I wanted to change the line you commented as follows
```suggestion
depends_on:MBEDTLS_THREADING_PTHREAD:MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE > MBEDTLS_CTR_DRBG_MAX_REQUEST + 1
```
but of course the Python script complains that this is an `Invalid dependency`. Do you have any suggestion on how I can express this dependency here?

gilles-peskine-arm replied:
> checking generate_test_code.py it seems to me that conditional depends_on should have the form:
> `<C_symbol> <operator> <numerical value>`

Indeed. That is annoying.

The only thing that `generate_test_code.py` does with `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE > MBEDTLS_CTR_DRBG_MAX_REQUEST + 1` is to inject it into an `#if` line. So we could actually relax this condition. For an entry in a `depends_on` clause, all `generate_test_code.py` wants to know is whether this is a bare symbol (turned into `defined(…)`), an exclamation point followed by a bare symbol (turned into `!defined(…)`), or something else (turned into just `(…)`).

We could in principle allow anything and let the C preprocessor complain if the expression is invalid. I think this would be too extreme because the feedback on errors (e.g. typing `FOO;BAR` instead of `FOO:BAR`) is pretty poor: you'd get an error when compiling the `.c` file, with a meaningless error line number in `main_test.function` (there's no easy way to emit a `#line` statement referring to the `.data` file). But we could at least allow a larger subset. For example:

* must contain a relation operator
* may not contain `;` or `,`
* probably shouldn't contain `?:` either, because the `:` would be interpreted as separating dependencies

I think “must only contain plausible characters in a preprocessor expression, and must contain a relation operator” would be a reasonable compromise. How about this (untested)?

```
CONDITION_REGEX = re.compile(
    r'(?!.*\b0[0-9])' # forbid 0ddd which might be accidentally octal or accidentally decimal
    r'(?=.*[!=]=|[<>]=)' # must contain a relation operator
    r'[\s\w!%&()*+-/<=>^|~]+' # plausible characters in an integer constant expression
, re.ASCII)
```

valeriosetti replied:
Actually I just changed `CONDITION_OPERATOR_REGEX` in `generate_test_code.py` so that now the 2nd element of the comparison can be a 'composition" of symbols, math operators and values. Unfortunately the change still falls into the `framework` subrepo, so I just added one more commit [there](https://github.com/Mbed-TLS/mbedtls-framework/pull/48/commits/fe07e3c9e0dc77191fd67aba57e88f1fd7ecf8e6).

valeriosetti commented on tests/include/test/psa_crypto_helpers.h:
```diff
@@ -461,4 +461,26 @@ uint64_t mbedtls_test_parse_binary_string(data_t *bin_string);
+/* Some helper macros to verify if MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE is
+ * large enough to contain an RSA key pair of the given size. This is meant to be
+ * used in test cases where MBEDTLS_PSA_STATIC_KEY_SLOTS is enabled. */
+#if defined(MBEDTLS_PSA_CRYPTO_C)
```
Post CI run comment: a reason for having `MBEDTLS_TEST_STATIC_KEY_SLOTS_SUPPORT_RSA_[2048/4096]` guarded by `MBEDTLS_PSA_CRYPTO_CLIENT` is that `PSA_KEY_EXPORT_RSA_KEY_PAIR_MAX_SIZE()` is not defined when there's no PSA included in the build. This would cause tests like `test_no_psa_crypto_full_cmake_asan` to fail during the build because of errors like this:

```
tests/include/test/psa_crypto_helpers.h:467:49: error: function-like macro 'PSA_KEY_EXPORT_RSA_KEY_PAIR_MAX_SIZE' is not defined
#if (MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE >= PSA_KEY_EXPORT_RSA_KEY_PAIR_MAX_SIZE(4096))
                                                ^
```

Therefore I think I will keep the `MBEDTLS_PSA_CRYPTO_CLIENT` guard for these symbols.

valeriosetti commented:
I think that failures on the InternalCI can be ignored as the OpenCI is fully green. Therefore I would say that the PR is ready for reviews again :)

gilles-peskine-arm reviewed (changes requested):
One macro shouldn't be in a public header, and a couple of test case dependencies don't seem right. Otherwise LGTM

gilles-peskine-arm reviewed (approved):
LGTM

gilles-peskine-arm reviewed (changes requested):
Sorry, I almost forgot: looks good to me except for that we need a changelog entry.

tom-cosgrove-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -3066,6 +3066,26 @@
+ * #MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE.
+ *
+ * \note This feature comes with a (potentially) higher RAM usage since:
+ *       - All the key slots are allocated no matter if they are used of not.
```
(nit) `or not`, not `of not`

tom-cosgrove-arm commented on include/mbedtls/mbedtls_config.h:
```diff
@@ -3867,6 +3887,19 @@
+ *
+ * Define the size (in bytes) of each static key buffer when
+ * #MBEDTLS_PSA_STATIC_KEY_SLOTS is set. If not
+ * explicitly defined then it's automatically guessed from available PSA keys
```
Rather than `guessed`, would `determined` be better? `Guess` doesn't really have the right connotations

gilles-peskine-arm replied:
I think guess is fine, given that we're using heuristics that we don't want to commit to.

tom-cosgrove-arm commented on tests/include/test/psa_crypto_helpers.h:
```diff
@@ -461,4 +461,43 @@ uint64_t mbedtls_test_parse_binary_string(data_t *bin_string);
+#endif
+
+/* Helper macro for the PK module to check whether MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE
+ * is large enough to contain 4096 bits RSA key pairs. Of course this check is only
```
(nit) `4096-bit RSA key pairs` rather than `4096 bits...`

tom-cosgrove-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -56,6 +75,68 @@ component_build_psa_crypto_spm () {
+    scripts/config.py set MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG
+    # Set calloc/free as null pointer functions. Calling them would crash
+    # the program so we can use this as a "sentinel" for being sure no module
+    # is making use of these functions in the library.
```
I like this - nice

tom-cosgrove-arm commented on tf-psa-crypto/core/psa_crypto_core.h:
```diff
@@ -155,7 +155,11 @@ typedef struct {
+        uint8_t data[MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE];
+#else /* MBEDTLS_PSA_STATIC_KEY_SLOTS */
         uint8_t *data;
+#endif /* MBEDTLS_PSA_STATIC_KEY_SLOTS */
```
Since the span of this `#if-else-endif` is so short, it's probably better for readability to remove the `MBEDTLS_PSA_STATIC_KEY_SLOTS` comments on the `#else` and `#endif` lines

tom-cosgrove-arm commented on tf-psa-crypto/include/psa/crypto_sizes.h:
```diff
@@ -1085,6 +1089,27 @@
+#elif defined(PSA_WANT_KEY_TYPE_DES)
+#define PSA_CIPHER_MAX_KEY_LENGTH       24u
+#else
+#define PSA_CIPHER_MAX_KEY_LENGTH       0u
```
Is there a test that creates a configuration with this value chosen? I can imagine problems if this is used to size an array somewhere...

tom-cosgrove-arm reviewed (commented):
A couple of minor nits, otherwise looks good

gilles-peskine-arm commented on ChangeLog.d/9302.txt:
```diff
@@ -0,0 +1,8 @@
+   * When the new compilation option MBEDTLS_PSA_STATIC_KEY_SLOTS is enabled,
+     key store uses static key slot buffers instead of dynamically allocated
+     ones.
+   * The size of each static key slot buffer is determined by
+     MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE. If not set, its default value
+     is set to accomodate for the largest PSA asymmetric/symmetric key enabled
+     in the build.
```
I don't think `MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE` is really a separate feature from `MBEDTLS_PSA_STATIC_KEY_SLOTS`. Rather, I think it should be mentioned in the entry for `MBEDTLS_PSA_STATIC_KEY_SLOTS`. E.g.

> (…) The buffer size is given by the ption MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE. By default, it accommodates the largest expected key.

tom-cosgrove-arm replied:
This reads like it's part of the previous bullet point, but fell out. If we want it as a separate bullet point, it should start something like "Add new config option MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE, which ..."

gilles-peskine-arm commented on ChangeLog.d/9302.txt:
```diff
@@ -0,0 +1,8 @@
+Features
+   * When the new compilation option MBEDTLS_PSA_STATIC_KEY_SLOTS is enabled,
+     key store uses static key slot buffers instead of dynamically allocated
+     ones.
```
I'm not sure this is clear to people who don't know the feature. It's saying that static key slots are static and not dynamic, which feels repetitive. Can we make the English description more informative? Hopefully work in that this means the static key store doesn't use malloc?

tom-cosgrove-arm replied:
How about something like

```
   * Add new configuration option MBEDTLS_PSA_STATIC_KEY_SLOTS, which
     pre-allocates storage for keys, enabling malloc-less use of key slots.
```

gilles-peskine-arm replied:
I like the concision, but “pre-allocates” is vague. How about “which makes the key store a statically allocated global array, enabling …”?

tom-cosgrove-arm replied:
Or just replace "pre-allocates" with "uses static"

```
   * Add new configuration option MBEDTLS_PSA_STATIC_KEY_SLOTS, which
     uses static storage for keys, enabling malloc-less use of key slots.
```

gilles-peskine-arm reviewed (changes requested):
I'd like some improvements to the changelog entry

gilles-peskine-arm reviewed (dismissed):
Thanks for the updates! Looks good to me now.

tom-cosgrove-arm commented on ChangeLog.d/9302.txt:
```diff
@@ -0,0 +1,8 @@
+Features
+   * When the new compilation option MBEDTLS_PSA_STATIC_KEY_SLOTS is enabled,
+     key store uses statically allocated key slot's buffers instead of
```
Sorry, under a "new features" section (which is where this will appear) I would expect more of "Added new ..." rather than "When ...". 

`statically allocated key slot's buffers` is clunky - the simplest way to fix that is `statically-allocated key slot buffers` (remove the `'s`)

gilles-peskine-arm replied:
We aren't consistent about present tense vs past tense. The entry for the dynamic key store starts with “When the new compilation option MBEDTLS_PSA_KEY_STORE_DYNAMIC is enabled”, and I think that's fine.

valeriosetti commented:
I updated the first sentence of the changelog with Tom's [proposal](https://github.com/Mbed-TLS/mbedtls/pull/9302#discussion_r1787926372).

I have a question. CI complains that there's a conflict on the `framework` reference and I suspect that this is due to the fact that the `development` reference was updated by some other PR while this one was being reviewed, so the change I'm implementing here is not a fast-forward for `framework`. What should I do? Do I need to rebase https://github.com/Mbed-TLS/mbedtls-framework/pull/48 on top of the `main` branch, or will this be "magically" fixed by github during the merge?

gilles-peskine-arm commented:
For the framework, the workflow has to be:

1. Merge the framework pull request, so that the required framework content is present on the `main` branch of the framework repository.
2. Update the PR in `mbedtls` to point to a commit on the framework's `main` branch.

We have to do it this way so that the framework submodule in `development` always points to a commit in the submodule's `main` branch and not in some random PR.

tom-cosgrove-arm commented on ChangeLog.d/9302.txt:
```diff
@@ -0,0 +1,6 @@
+   * Added new configuration option MBEDTLS_PSA_STATIC_KEY_SLOTS, which
+     uses static storage for keys, enabling malloc-less use of key slots.
+     The size of each buffer is given by the option
+     MBEDTLS_PSA_STATIC_KEY_SLOT_BUFFER_SIZE. By default it accomodates the
```
(spelling) sorry, missed it before, but this should be `accommodates`.

Other than that, this all looks good to me

tom-cosgrove-arm reviewed (dismissed):
LGTM

gilles-peskine-arm reviewed (approved):
LGTM

valeriosetti commented:
@gilles-peskine-arm @tom-cosgrove-arm sorry for dismissing your approval but I had to rebase this PR and add a new commit updating the `framework` reference due to CI failures reported in the [backport](https://github.com/Mbed-TLS/mbedtls/pull/9448#issuecomment-2401824412).

I suggest to wait for this PR's CI to return green before taking another look. Btw a part from the last commit nothing else was changed in the code

tom-cosgrove-arm reviewed (dismissed):
Reconfirming - no change to content (apart from `framework` pointer)

tom-cosgrove-arm commented:
The CI pr-merge failed on the ABI-API check, which is expected due to the framework change (`mbedtls_version_get_foo()` removed)

gilles-peskine-arm reviewed (dismissed):
LGTM

bensze01 commented:
@valeriosetti Please update the framework submodule to d9a70c758a5a511020240bbcc86f6d647d64fe96

bensze01 commented:
Hmm, Github still shows a conflict on the framework submodule. I guess the commits that update the submodule need to be rebased / dropped (except the most recent one)

gilles-peskine-arm reviewed (dismissed):
LGTM after rebase from 9a07d36adb23f62814bae55ff8f60fe89b4f634b to update the framework

valeriosetti commented:
Since there was (**another**) conflict with `framework`, but https://github.com/Mbed-TLS/mbedtls-framework/pull/48 has been merged, I decided to rebase this PR on `development` and remove _all_ the commits which were updating the `framework` reference. The rest of the code is untouched, so the PR is _ready_ for reviews again.

valeriosetti commented:
Errata on my last comment: I had to add a new commit to the series in order to make (hopefully) the CI happy. The reason is that I encountered issues in `analyze_outcomes.py` in Coverage tests: it was expecting that `!PSA_WANT_ALG_TLS12_PRF` are never executed, but this is not correct for the new test component that I added in this PR.

valeriosetti commented:
OpenCI is green. Just let me know what you think about the last commit so that I can backport it to #9448 and keep the 2 PRs aligned.

gilles-peskine-arm reviewed (approved):
Looks good to me at  for:

* rebase from my previous approval of ad9ef1d4adb95141f5b89a427fcd76f4e791a605 at https://github.com/Mbed-TLS/mbedtls/pull/9302#pullrequestreview-2372237834
* The framework update is already in merged `development` so no framework update is needed now
* The extra commit to update `analyze_outcomes.py`.

Since this resolves #9577 in `development` and the issue is not present in 3.6, we need to close that issue.
